include/igl/generate_sparse_marching_cubes_grid.cpp

```diff
@@ -0,0 +1,152 @@
+#include "generate_sparse_marching_cubes_grid.h"
+
+#include <unordered_map>
+#include <array>
+#include <vector>
+
+
+struct RowVector3iHash {
+    std::size_t operator()(const Eigen::RowVector3i& key) const {
+        std::size_t seed = 0;
+        for (int i = 0; i < 3; i++) {
+            seed ^= key[i] + 0x9e3779b9 + (seed<<6) + (seed>>2); // Copied from boost::hash_combine
+        }
+        return std::hash<std::size_t>()(seed);
+    }
+};
+
+template <typename T>
+int sgn(T val) {
+    return (T(0) < val) - (val < T(0));
+}
+
+
+template<class DerivedM, class DerivedR>
+int matrix_push_back(Eigen::PlainObjectBase<DerivedM>& mat, const Eigen::PlainObjectBase<DerivedR>& row, int n) {
+    if (mat.rows() == n) {
+        mat.conservativeResize(2*n+1, mat.cols());
+    }
+
+    mat.row(n) = row;
+    return n + 1;
+}
+
+
+template<class DerivedM>
+int vector_push_back(Eigen::PlainObjectBase<DerivedM>& mat, typename Eigen::PlainObjectBase<DerivedM>::Scalar row, int n) {
+    if (mat.rows() == n) {
+        mat.conservativeResize(2*n+1);
+    }
+
+    mat[n] = row;
+    return n + 1;
+}
+
+
+
+template <typename DerivedS, typename DerivedV, typename DerivedI>
+void igl::generate_sparse_marching_cubes_grid(const Eigen::RowVector3d& p0, std::function<double(const Eigen::RowVector3d&)> scalarFunc, double eps,
```

Comment:
Daniele had a memory corruption the other day because of this. He was assigning the const reference to another const reference in a copy constructor. It looks like the compiler made a copy of the memory and the reference pointed to garbage when the constructor finished.